src/main/java/gregtech/common/covers/CoverFluidRegulator.java

```diff
@@ -0,0 +1,248 @@
+            default: return 0;
+        }
+    }*/
+
+    @Override
+    protected int doTransferFluidsInternal(IFluidHandler myFluidHandler, IFluidHandler fluidHandler, int transferLimit) {
+        IFluidHandler sourceHandler;
+        IFluidHandler destHandler;
+
+        if (pumpMode == PumpMode.IMPORT) {
+            sourceHandler = fluidHandler;
+            destHandler = myFluidHandler;
+        } else if (pumpMode == PumpMode.EXPORT) {
+            sourceHandler = myFluidHandler;
+            destHandler = fluidHandler;
+        } else {
+            return 0;
+        }
+        switch (transferMode) {
+            case TRANSFER_ANY: return GTFluidUtils.transferFluids(sourceHandler, destHandler, transferLimit, fluidFilter::testFluidStack);
+            case KEEP_EXACT: return doKeepExact(transferLimit, sourceHandler, destHandler, fluidFilter::testFluidStack, this.keepAmount);
+            case TRANSFER_EXACT: return doTransferExact(transferLimit, sourceHandler, destHandler, fluidFilter::testFluidStack, this.supplyAmount);
+        }
+        return 0;
+    }
+
+    protected int doTransferExact(int transferLimit, IFluidHandler sourceHandler, IFluidHandler destHandler, Predicate<FluidStack> fluidFilter, int supplyAmount) {
+        int fluidLeftToTransfer = transferLimit;
+        for (IFluidTankProperties tankProperties : sourceHandler.getTankProperties()) {
+            if (fluidLeftToTransfer < supplyAmount)
+                break;
+            FluidStack sourceFluid = tankProperties.getContents();
+            if (sourceFluid == null || sourceFluid.amount == 0 || !fluidFilter.test(sourceFluid)) continue;
+            sourceFluid.amount = supplyAmount;
+            if (GTFluidUtils.transferExactFluidStack(sourceHandler, destHandler, sourceFluid.copy())) {
+                fluidLeftToTransfer -= sourceFluid.amount;
+            }
+            if (fluidLeftToTransfer == 0) break;
+        }
+        return transferLimit - fluidLeftToTransfer;
+    }
+
+    protected int doKeepExact(int transferLimit, IFluidHandler sourceHandler, IFluidHandler destHandler, Predicate<FluidStack> fluidFilter, int keepAmount) {
+        int fluidLeftToTransfer = transferLimit;
+        for (IFluidTankProperties tankProperties : sourceHandler.getTankProperties()) {
+            FluidStack sourceFluid = tankProperties.getContents();
+            if (sourceFluid == null || sourceFluid.amount == 0 || !fluidFilter.test(sourceFluid)) continue;
+            sourceFluid.amount = keepAmount;
+            FluidStack destFluid = destHandler.drain(sourceFluid, false);
+            int amountToDrainAndFill;
+            //no fluid in destination
+            if (destFluid == null) {
+                amountToDrainAndFill = Math.min(keepAmount, fluidLeftToTransfer);
+            //if the amount of fluid in the destination is sufficient or the destinations fluid isnt equal to the sources
+            //how to check if destHandler is full?
```

Comment:
To check if the destination tank is full, you could check the tank properties to the destination handler and get the capacity and contents and compare them.

Reply:
I think that was an old comment that I made for myself. In my testing I have run into no issues without including this check. I resolved the issue of the pumps accidentally being fluid regulators. Regarding the text not being centered, this stems from `AdvancedTextWidget::drawInBackground`. The other fields are `SimpleTextWidget` which centers the text in that method. I don't feel this falls in the scope of this PR and would be better suited in a separate one.

Reply:
I agree that fixing this widget is not in the scope of this PR. Would you mind making an issue for it through, so it is not forgotten?

Reply:
Done